typescript/connections.md

```diff
@@ -0,0 +1,209 @@
+# SDK Connection Configuration
+
+## Problem statement
+
+The current API for connecting a Worker to the server is via a `Core` singleton.
+This singleton can only connect to one namespace per process and, by looking at our basic samples,
+it not immediately obvious how to configure the server address for a given Worker.
+
+```ts
+// worker.ts
+await Core.install({ serverOptions: { address, namespace } });
+// Worker uses a connection created by the Core singleton
+const worker = await Worker.create({ taskQueue });
+```
+
+## Proposals
+
+### Explicit `Connection` objects
```

Comment:
After discussion and re-review, I am afraid I begrudgingly support this option as best.

I think it is very important to have these goals:

* Connection sharing should be default, not opt-in
* That you are sharing a connection should be clear
  * Worker creation therefore must be a two-step process since they need to be created from a common base of shared something
* Prefer to avoid global state

I think asking them to create a connection, just as if gRPC required a channel to be created explicitly (which I know it does not in all gRPC impls, but it does in many) is the best way forward. It does mean an extra step. I'd also add:

* I would like to just call it `WorkerConnection` or something similar I dunno. `NativeConnection` reads a bit rough to me.
* I would prefer `WorkflowClient` constructor just accept the connection instead of `connection.service`. It just feels more right to create a client from a connection than from a connection service (you can type the param as `connection: { service: WorkflowService }` if you must but I prefer accepting exactly what's returned from `Connection.create` since that's easier to understand where it came from for API browsing).

Reply:
> That you are sharing a connection should be clear

Why is this important for majority of devs to know?

Reply:
I tend to agree with @cretz, and suggest that we go with "Explicit Connection objects".
That approach is explicit, future proof for when we want to support different connection types, and flexible (for example we could extend this API to create implicit connections when given options).

Reply:
I'd like to call the connection `NativeConnection` in case we want to use it in the client in the future or we support running workers with non-native connections.

I would change the `WorkflowClient` constructor but I don't want to tie that change to this proposal because I might take it a step further and unify the 2 clients into a single `Client` in the future.

Reply:
> Why is this important for majority of devs to know?

Clarity of when connections are being made is helpful like any other form of clarity. We need a really good reason to argue against clarity. It's a tradeoff. The question is whether it's more important for them not to know.

Reply:
Another reason this approach is good: dependency injection. In frameworks like Nest.js, the most natural thing is to create a connection within the DI context, then inject it into workers (or wherever else) upon creation. 

Even without explicit connection creation, you'd still want to create a `ConnectionOptions` object within DI context and inject _that_ into the workers. So a 2-step creation is a given when using DI, and it's much clearer to just create the connection directly.

Regarding the confusion about when to use each connection type, it seems like that could be solved with more specific types for `Worker.create()` and `new WorkflowClient()`.

Reply:
> Why is it important to hide it from all devs?

Simplicity / lower mental load for the majority for whom connection constructing doesn't matter. And it wouldn't have to be completely hidden—`Worker.create` API doc can say "Creates a gRPC connection to the server, or reuses an existing connection if another Worker has previously been created with the same `connectionOptions`".

> That approach is explicit, future proof for when we want to support different connection types, and flexible (for example we could extend this API to create implicit connections when given options).

I'd be happy with this extension, and using it for most of samples. In which case the above API doc could be in `WorkerOptions.connectionOptions` rather than `Worker.create`.

Reply:
Seems I'm a little bit to the show, but for what it it's worth, here's my two cents...

First of all, I think that being explicit on connection instantiation will be wonderful from a DI point of view.

I'm however concerned that, in its current form, this proposal exposes the user to some lower-level aspects of how Temporal TS works. This might not be an issue for users that are already comfortable with Temporal, but may be a turn off for new users.

Here are some of the irritants I see:
1) as a new user, I need to know that there are two types of connection object, and I have to know which one to provide when creating a client or setting up a worker... And even worst, this knowledge is likely to change at some point in time which means that code examples I see on the internet may not work or may be non optimal in the future.
2) as a user, I need to understand the cost associated with the connection object and therefore, that I am in charge of reusing the connection object. Simply having a connection object doesn't make it explicit that I am meant to reuse it and that doing so is desirable.
3) Knowing that I can reuse the connection now makes me wonder about the lifecycle of objects I create in association with that connection. Am I expected to have some kind of global connection that I expose to my whole application, and then create a client attached to it whenever I need one, and then forget about it? Is there a cost associated ok to have multiple clients attached to one connection? Do I need to explicitly dispose of client/workers? How do I dispose of a connection, and what will be the consequence on worker/clients I created from it? All of this can be clarified with good documentation, but that means that I, as a user, have to take the time to read that doc...

Reply:
I've been putting off this reply for a while, sorry for that.

You raise some valid points as usual @mjameswh.

> as a new user, I need to know that there are two types of connection object, and I have to know which one to provide when creating a client or setting up a worker... And even worst, this knowledge is likely to change at some point in time which means that code examples I see on the internet may not work or may be non optimal in the future.

I see us at some point soon adding a method to create workers and clients from connection options instead of passing around connection objects.

 > as a user, I need to understand the cost associated with the connection object and therefore, that I am in charge of reusing the connection object. Simply having a connection object doesn't make it explicit that I am meant to reuse it and that doing so is desirable.

I agree, this would be solved by accepting connection options and defaulting to implicitly sharing connections.
That being said, I'm not convinced this is the right approach. In general I'd like to avoid implicit global state and not sure how to solve this differently.

> Knowing that I can reuse the connection now makes me wonder about the lifecycle of objects I create in association with that connection. Am I expected to have some kind of global connection that I expose to my whole application, and then create a client attached to it whenever I need one, and then forget about it? Is there a cost associated ok to have multiple clients attached to one connection? Do I need to explicitly dispose of client/workers? How do I dispose of a connection, and what will be the consequence on worker/clients I created from it? All of this can be clarified with good documentation, but that means that I, as a user, have to take the time to read that doc...

Good point, the answer is yes, you're supposed to keep a single connection (per role - starter / worker) and reuse that throughout the app. Client objects are relatively cheap so those don't have to be reused but typically we suggest reusing them.
As a rule of thumb, I'm trying to keep "expensive" constructors in async factory functions (e.g. `create`), that should be a guiding factor for users to decide what to reuse.
While this answer might not be completely satisfactory, we'll make up for this complexity with good samples and docs.


I'm open to other suggestions but I feel like we took the most forwards compatible and flexible path without introducing implicitness and global state.